### PR TITLE
Start cache in non leader elected instances

### LINKED ISSUE #165: Failed to wait for cert-rotator caches to sync in non-leader elected instances
In the case the cert controller is added to a non-leader manager, i.e., with `CertRotator.RequireLeaderElection` set to false, it fails with the following error message:

```
{"level":"error","ts":"2023-12-12T18:04:44.726776367Z","caller":"controller/controller.go:203","msg":"Could not wait for Cache to sync","controller":"cert-rotator","error":"failed to wait for cert-rotator caches to sync: timed out waiting for cache to be synced for Kind *v1.Secret","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.1\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:203\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:208\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:234\nsigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func1\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/manager/runnable_group.go:223"}
```

### PR DESCRIPTION
**Pull request by astefanutti**
Fixes #165.

This PR makes sure the cache is started in non-leader elected instances.

### CODE REVIEW

File: pkg/rotator/rotator.go [hunk 1]
```diff
@@ -709,7 +719,7 @@ func addController(mgr manager.Manager, r *ReconcileWH) error {
 		}
 	}
 
-	return nil
+	return mgr.Add(controllerWrapper{c, r.needLeaderElection})
```
Comment (alculquicondor) [inline pkg/rotator/rotator.go:722]: Is this change part of the fix too?

File: pkg/rotator/rotator.go [hunk 2]
```diff
@@ -191,13 +202,15 @@ func addNamespacedCache(mgr manager.Manager, namespace string) (cache.Cache, err
 	if err != nil {
 		return nil, err
 	}
-	if err := mgr.Add(c); err != nil {
+	// Wrapping the cache to make sure it's also started when the manager
+	// hasn't been leader elected and CertRotator.RequireLeaderElection is false.
+	if err := mgr.Add(&cacheWrapper{Cache: c}); err != nil {
```
Comment (alculquicondor) [inline pkg/rotator/rotator.go:207]: how would someone set up this boolean?

Comment (astefanutti) [reply pkg/rotator/rotator.go]: Ah good catch, I had local changes. Updated.

File: pkg/rotator/rotator.go (hunk 1 above)
Comment (astefanutti) [reply pkg/rotator/rotator.go:722]: Yes it is.

**Comment by alculquicondor** [comment]
/lgtm
?

**Comment by astefanutti** [comment]
@maxsmythe @sozercan Could you please take a look?

**Review by maxsmythe — APPROVED** [review]
This fixes a problem where the manager participates in leader election, but the cert rotator is not wired to require leader election, is that correct?

LGTM, thank you for the fix!

**Comment by astefanutti** [comment]
> This fixes a problem where the manager participates in leader election, but the cert rotator is not wired to require leader election, is that correct?

@maxsmythe that’s exactly that. Thanks a lot for the quick review!

**Comment by alculquicondor** [comment]
Anything else pending to merge?

**Comment by astefanutti** [comment]
@sozercan @ritazh do you want to have a look before merging?

File: pkg/rotator/rotator.go [hunk 3]
```diff
@@ -191,13 +202,15 @@ func addNamespacedCache(mgr manager.Manager, namespace string) (cache.Cache, err
 	if err != nil {
 		return nil, err
 	}
-	if err := mgr.Add(c); err != nil {
+	// Wrapping the cache to make sure it's also started when the manager
+	// hasn't been leader elected and CertRotator.RequireLeaderElection is false.
+	if err := mgr.Add(&cacheWrapper{Cache: c, needLeaderElection: cr.RequireLeaderElection}); err != nil {
```
Comment (ritazh) [inline pkg/rotator/rotator.go:207]: where do we check if CertRotator.RequireLeaderElection is false?

Comment (astefanutti) [reply pkg/rotator/rotator.go:207]: It is ultimately checked by controller-runtime, that calls the `NeedLeaderElection` method here:

https://github.com/kubernetes-sigs/controller-runtime/blob/1b80b9629af8ec55ae200e2c649a9603533eed20/pkg/manager/runnable_group.go#L66

This is the same mechanism used to control how the other runnables, i.e., `CertRotator` and `ReconcileWH`, start w.r.t. leader election.

**Review by ritazh — APPROVED** [review]
LGTM

Thanks for the PR! @astefanutti

**Comment by astefanutti** [comment]
@ritazh @maxsmythe thanks a lot for the review!

Do you have plan to release a new version anytime soon? We would kindly appreciate a new one if possible 🙏🏼.

**Comment by maxsmythe** [comment]
Published a patch release v0.10.1

**Comment by astefanutti** [comment]
> Published a patch release v0.10.1

@maxsmythe awesome, thanks a lot for the quick release!

**Pull request merged**